adapters/repos/db/queue/worker.go

```diff
@@ -102,67 +99,60 @@ func (w *Worker) do(batch *Batch) (err error) {
 			return nil // all tasks succeeded
 		}
 
-		if attempts >= maxRetry {
+		hasPermanentErrs := hasPermanentErrors(errs)
+		if hasPermanentErrs {
+			w.logger.WithError(errors.Join(errs...)).
+				WithField("failed", len(failed)).Error("permanent errors detected, discarding batch")
+			return nil
+		}
+
+		hasTransientErrs := hasTransientErrors(errs)
+		if hasTransientErrs {
+			retryIn := w.calculateBackoff(attempts)
 			w.logger.
 				WithError(errors.Join(errs...)).
 				WithField("failed", len(failed)).
 				WithField("attempts", attempts).
-				Error("failed to process task, discarding")
-			return nil
-		}
-
-		w.logger.
-			WithError(errors.Join(errs...)).
-			WithField("failed", len(failed)).
-			WithField("attempts", attempts).
-			Infof("failed to process task, retrying in %s", w.retryInterval.String())
-		attempts++
-
-		t := time.NewTimer(w.retryInterval)
-		select {
-		case <-batch.Ctx.Done():
-			// drain the timer
-			if !t.Stop() {
-				<-t.C
+				WithField("retry_in", retryIn).
+				Warnf("transient errors detected, retrying batch in %s", retryIn)
+
+			attempts++
+			retryTimer := time.NewTimer(retryIn)
+			select {
+			case <-batch.Ctx.Done():
+				if !retryTimer.Stop() {
+					<-retryTimer.C
+				}
+				return batch.Ctx.Err()
+			case <-retryTimer.C:
+				// try again
 			}
-
-			return batch.Ctx.Err()
-		case <-t.C:
 		}
 	}
 }
 
-func (w *Worker) executeTaskRetryOnTransient(ctx context.Context, t Task) error {
-	linearBackoff := backoff.NewExponentialBackOff()
-	linearBackoff.InitialInterval = w.retryInterval
-	linearBackoff.Multiplier = 1.0 // Linear growth (no exponential)
-	linearBackoff.MaxInterval = 3 * time.Second
-	linearBackoff.MaxElapsedTime = 0 // Retry forever for transient errors
-
-	backoffWithContext := backoff.WithContext(linearBackoff, ctx)
-
-	attempt := 1
-	return backoff.Retry(func() error {
-		err := t.Execute(ctx)
-		if err == nil {
-			// no error, fast path return
-			return nil
-		}
+func (w *Worker) calculateBackoff(attempts int) time.Duration {
+	backoff := time.Duration(1<<uint(attempts-1)) * time.Second
```

Comment:
Can we add a comment here just to highlight the bit shifting at play for clarity when reading the code at a glance?